File: parboiled-scala/src/test/scala/org/parboiled/scala/SimpleTest.scala
/*
 * Copyright (C) 2009-2011 Mathias Doenitz
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.parboiled.scala

import org.testng.annotations.Test
import org.scalatest.testng.TestNGSuite
import org.testng.Assert.assertEquals
import org.parboiled.matchers.Matcher
import org.parboiled.support.ToStringFormatter
import org.parboiled.trees.GraphUtils
import testing.ParboiledTest

class SimpleTest extends ParboiledTest with TestNGSuite {

  class SimpleParser extends Parser {
    def Clause = rule {Digit ~ ClauseRest ~ EOI}

    def ClauseRest = rule {zeroOrMore(Operator ~ Digit) ~ anyOf("abcd")}

    def Operator = rule {"+" | "-"}

    def Digit = rule {"0" - "9"}
  }

  val parser = new SimpleParser() {
    override val buildParseTree = true
  }

  @Test
  def testRuleTreeConstruction() {
    val rule = parser.Clause
    assertEquals(GraphUtils.printTree(rule.matcher, new ToStringFormatter[Matcher]),
       """|Clause
          |  Digit
          |  ClauseRest
          |    ZeroOrMore
          |      Sequence
          |        Operator
          |          '+'
          |          '-'
          |        Digit
          |    [abcd]
          |  EOI
          |""".stripMargin);
  }

  @Test
  def testSimpleParse() {
    parse(ReportingParseRunner(parser.Clause), "1+2a") {
      assertEquals(parseTree,
         """|[Clause] '1+2a'
            |  [Digit] '1'
            |  [ClauseRest] '+2a'
            |    [ZeroOrMore] '+2'
            |      [Sequence] '+2'
            |        [Operator] '+'
            |          ['+'] '+'
            |        [Digit] '2'
            |    [[abcd]] 'a'
            |  [EOI]
            |""".stripMargin)
    }
  }

}

File: parboiled-java/src/test/java/org/parboiled/SimpleTest.java
/*
 * Copyright (C) 2009-2011 Mathias Doenitz
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.parboiled;

import org.parboiled.annotations.BuildParseTree;
import org.parboiled.test.TestNgParboiledTest;
import org.testng.annotations.Test;

public class SimpleTest extends TestNgParboiledTest<Object> {

    @BuildParseTree
    static class Parser extends BaseParser<Object> {

        public Rule Clause() {
            return Sequence(Digit(), Operator(), Digit(), AnyOf("abcd"), OneOrMore(NoneOf("abcd")), EOI);
        }

        public Rule Operator() {
            return FirstOf(Ch('+'), '-');
        }

        public Rule Digit() {
            return CharRange('0', '9');
        }

    }

    @Test
    public void test() {
        Parser parser = Parboiled.createParser(Parser.class);
        Rule rule = parser.Clause();
        test(rule, "1+5bx")
                .hasNoErrors()
                .hasParseTree("" +
                        "[Clause] '1+5bx'\n" +
                        "  [Digit] '1'\n" +
                        "  [Operator] '+'\n" +
                        "    ['+'] '+'\n" +
                        "  [Digit] '5'\n" +
                        "  [[abcd]] 'b'\n" +
                        "  [OneOrMore] 'x'\n" +
                        "    [![abcdEOI]] 'x'\n" +
                        "  [EOI]\n");
    }

}
